arches/app/datatypes/url.py

```diff
@@ -82,6 +83,14 @@ def validate(self, value, row_number=None, source=None, node=None, nodeid=None,
                 title = _("Invalid HTTP/HTTPS URL")
                 error_message = self.create_error_message(value, source, row_number, message, title)
                 errors.append(error_message)
+            
+            # raise error if label added without URL (#10592)
+            if value.get("url_label") and not value.get("url"):
+                message = _("URL label cannot be saved without a URL")
+                title = _("No URL added")
+                error_message = self.create_error_message(value, source, row_number, message, title)
+                errors.append(error_message)
```

Comment:
@chiatt mentioned at the developer meeting last month (btw, great to meet you there!) that we're trying to improve the test coverage of the python backend. There's a test file for the datatype validation methods [here](https://github.com/archesproject/arches/blob/0c89e7d117f4c639f1ce706f3dfaa82462fb1ca4/tests/utils/datatype_tests.py#L4).

Would you be able to add a `URLDataTypeTests.test_validate()` method to cover your changes here?

Reply:
Great to meet you there too! 

I agree and was thinking about how this PR needed some tests during that presentation at the developer meetings.  I'll start working on writing some.

Reply:
I've written some tests for the url datatype in `datatype_tests.py` under the class name `URLDataTypeTests`, testing validation and pre_tile_save.  They are in the latest commit [685a5c6](https://github.com/archesproject/arches/pull/10642/commits/685a5c682bbac3614a4d44a789768440c4f755d4).

This was my first time writing django tests so please let me know if things need changing/adding @jacobtylerwalls @chiatt

Reply:
In the latest commit ([fd3c389](https://github.com/archesproject/arches/pull/10642/commits/fd3c3890bd0a5a405b1e287d892f9185f7de1baa)) I changed the line https://github.com/archesproject/arches/blob/0c89e7d117f4c639f1ce706f3dfaa82462fb1ca4/arches/app/datatypes/url.py#L75 to https://github.com/archesproject/arches/blob/fd3c3890bd0a5a405b1e287d892f9185f7de1baa/arches/app/datatypes/url.py#L76
This was because the test for a URL with an empty string and no label (e.g. `{"url": "", "url_label": "Label"}` was showing two errors, one for the fact a label couldn't be saved without a URL, and a second for the URL being invalid http/https.  This seemed wrong to me, as an empty string shoudn't be tested as a valid URL just because it isn't `None`.  Thoughts?

Reply:
I suppose the problem with that is that in the case of `{'url': '', 'url_label': ''}`, which you can have if you type into the fields and then backspace your entries before clicking save, the validation will now pass (when it shouldn't).

I think having both errors isn't hurting anything. The UI in the resource editor only presents the more useful error about the invalid URL.

Happy to hear @chiatt's thoughts too.

Reply:
Suggest adding `{'url': '', 'url_label': ''}` as a test case once we know how it should behave.

Reply:
It seems like what we need in the case of `{'url': '', 'url_label': ''}` is a `clean` method to return the value back to `None` on save. Perhaps something like:

```
def clean(self, tile, nodeid):
    if (data := tile.data[nodeid]):
        try:
            if not any([val.strip() for val in data.values()]):
                tile.data[nodeid] = None
        except Exception:
            pass # Let self.validate handle malformed data
```

Reply:
Thanks for this function @chiatt , it works well!  I've added it to the latest commit ([60c1af4](https://github.com/archesproject/arches/pull/10642/commits/60c1af4d61441bf1517777dc0ba9999a1260ec52)) and added the test_clean function in datatype_tests.py to test that None results from both empty URL and label values.